File: GFG/sorting/mergeSort.cpp
//The Merge Sort algorithm is a sorting algorithm that is based 
//on the Divide and Conquer paradigm.

//step 1: start
//step 2: declare array and left, right, mid variable
//step 3: perform merge function.
//    if left > right
//        return
//    mid= (left+right)/2
//    mergesort(array, left, mid)
//    mergesort(array, mid+1, right)
//    merge(array, left, mid, right)
//step 4: Stop

/* C program for Merge Sort */
#include <stdio.h>
#include <stdlib.h>
#include <vector>
using namespace std;

// Merges two subarrays of arr[].
// First subarray is arr[l..m]
// Second subarray is arr[m+1..r]
void merge(int arr[], int l, int m, int r)
{
	int i, j, k;
	int n1 = m - l + 1;
	int n2 = r - m;

	vector<int> L(n1), R(n2);

	/* Copy data to temp arrays L[] and R[] */
	for (i = 0; i < n1; i++)
		L[i] = arr[l + i];
	for (j = 0; j < n2; j++)
		R[j] = arr[m + 1 + j];

	/* Merge the temp arrays back into arr[l..r]*/
	i = 0; // Initial index of first subarray
	j = 0; // Initial index of second subarray
	k = l; // Initial index of merged subarray
	while (i < n1 && j < n2) {
		if (L[i] <= R[j]) {
			arr[k] = L[i];
			i++;
		}
		else {
			arr[k] = R[j];
			j++;
		}
		k++;
	}

	/* Copy the remaining elements of L[], if there
	are any */
	while (i < n1) {
		arr[k] = L[i];
		i++;
		k++;
	}

	/* Copy the remaining elements of R[], if there
	are any */
	while (j < n2) {
		arr[k] = R[j];
		j++;
		k++;
	}
}

/* l is for left index and r is right index of the
sub-array of arr to be sorted */
void mergeSort(int arr[], int l, int r)
{
	if (l < r) {
		// Same as (l+r)/2, but avoids overflow for
		// large l and h
		int m = l + (r - l) / 2;

		// Sort first and second halves
		mergeSort(arr, l, m);
		mergeSort(arr, m + 1, r);

		merge(arr, l, m, r);
	}
}

/* UTILITY FUNCTIONS */
/* Function to print an array */
void printArray(int A[], int size)
{
	int i;
	for (i = 0; i < size; i++)
		printf("%d ", A[i]);
	printf("\n");
}

/* Driver code */
int main()
{
	int arr[] = { 12, 11, 13, 5, 6, 7 };
	int arr_size = sizeof(arr) / sizeof(arr[0]);

	printf("Given array is \n");
	printArray(arr, arr_size);

	mergeSort(arr, 0, arr_size - 1);

	printf("\nSorted array is \n");
	printArray(arr, arr_size);
	return 0;
}

//T(n) = 2T(n/2) + θ(n)
//Time Complexity: O(N log(N))

File: GFG/sorting/MergeSort.java
import java.util.ArrayList;
import java.util.Scanner;

public class MergeSort {
    public static void merge(ArrayList<Integer> arr, Integer l, Integer mid, Integer r){
        ArrayList<Integer> temp1 = new ArrayList<Integer>();
        ArrayList<Integer> temp2 = new ArrayList<Integer>();

        for(Integer i = l; i <= mid; i++){
            temp1.add(arr.get(i));
        }
        for(Integer i = mid+1; i <= r; i++){
            temp2.add(arr.get(i));
        }
        int i = 0, j = 0, k = l;
        while(i<temp1.size() && j < temp2.size()){
            if(temp1.get(i) <= temp2.get(j)){
                arr.set(k,temp1.get(i));
                i++;
            }else{
                arr.set(k,temp2.get(j));
                j++;
            }
            k++;
        }
        while(i<temp1.size()){
            arr.set(k,temp1.get(i));
            i++;
            k++;
        }
        while(j<temp2.size()){
            arr.set(k, temp2.get(j));
            j++;
            k++;
        }
    }

    public ArrayList<Integer> mergesort(ArrayList<Integer> arr, Integer l, Integer r) {
        if (l < r) {
            var mid = (l + r) / 2;
            mergesort(arr, l, mid);
            mergesort(arr, mid + 1, r);
            merge(arr, l, mid, r);
        }
        return arr;
    }

    // Method to display array
    public static void printArray(ArrayList<Integer> arr) {
        for (int i = 0; i < arr.size(); i++) {
            System.out.print(arr.get(i) + " ");
        }
        System.out.println();
    }

    // Main method with user input
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        MergeSort ms = new MergeSort();
        
        System.out.print("Enter the number of elements: ");
        int n = scanner.nextInt();
        
        ArrayList<Integer> arr = new ArrayList<>();
        
        System.out.println("Enter " + n + " integers:");
        for (int i = 0; i < n; i++) {
            int element = scanner.nextInt();
            arr.add(element);
        }
        
        System.out.println("\nOriginal array:");
        printArray(arr);
        
        // Sort the array
        ms.mergesort(arr, 0, arr.size() - 1);
        
        System.out.println("\nSorted array:");
        printArray(arr);
        
        scanner.close();
    }
}
